Ho oh and lugia rework

bellarch opened the pull request:
Ho-oh's new sig ability, Rainbow Trail, cuts its Fire-type on entry and adds the type of each move it uses.
Lugia's new sig ability, Catastrophic, makes its moves always hit all other battlers and always invoke their random effects.
Ho-oh's new sig move, Golden Rebirth, is a 1 PP move that revives a fainted Pokemon in your party at half health, switches it into battle, and gives it +1 to all stats at the end of the turn.
Lugia's new sig move, Silver Gale, is a 1 PP move that clears all hazards, puts all hazards on your side of the field on your opponent's, then forces all affected battlers to switch at the end of the turn.
Implemented a new method to handle switching Pokemon selected by an external input rather than by either a player choice or at random.

bellarch committed:
Lugia and Ho-oh sig abilities + progress on ho-oh sig move

Catastrophic Force makes moves become spread+ and activate random effects. Rainbow Trail cuts Ho-oh's Fire type and makes it gain the types of moves it uses.

bellarch committed:
Implement Lugia sig move

Sig move clears opponent's hazards, then puts hazards on the user's side on the opponent's, then forces the opponent to switch randomly.
+ progress on Ho-oh sig move

bellarch committed:
fixing find/replace error

bellarch committed:
Implemented Ho-oh signature move

Blazing Rebirth revives a Pokemon at half HP, switches it into battle, and raises its stats by 1.
+ a new method for switching Pokemon that allows for a forced switch to a particular Pokemon has been created

bellarch committed:
changed the name of ho-oh's sig move

bellarch committed:
Finalizing text and tweaking code

in particular made sure Rainbow Trails could still add the Fire-type to Ho-oh and correctly ensured Golden Rebirth should fail if trapped

bellarch committed:
Merge pull request #1 from bellarch/development

bellarch committed:
Update trainers.txt

bellarch committed:
text adjustments

bellarch committed:
Merge pull request #2 from bellarch/development

merge

bellarch committed:
Update abilities_new.txt

bellarch committed:
Merge pull request #3 from bellarch/development

Development

bellarch committed:
cleaning up the trapping condition

bellarch committed:
Merge branch 'ho-oh-and-lugia-rework' of https://github.com/bellarch/Pokemon-Tectonic into ho-oh-and-lugia-rework

bellarch committed:
Adjusted Golden Rebirth's animations to work better

bellarch committed:
finalizing golden rebirth animation

bellarch committed:
Merge branch 'ho-oh-and-lugia-rework' into development

bellarch committed:
Merge pull request #4 from bellarch/development

Development

bellarch committed:
fixed an oopsie

zinniagamedev commented on PBS/abilities_new.txt:
```diff
@@ -1684,6 +1688,11 @@ Description = When the Pokémon is rampaging, it can choose the rampage moves it
+[RAINBOWTRAIL]
+Name = Rainbow Trail
+Description = Loses its Fire type. Whenever it uses a move, it gains that move's type.
+Flags = Immutable
```
Don't think this needs to be immutable.

bellarch replied:
The pinned message said immutable; let me double-check to see if the reason was discussed

bellarch replied:
Was briefly mentioned that people thought it should maybe be immutable (maybe to keep the lack of Fire type?) and I think I just incorporated it into the summary. Was not discussed in much detail. It does make sense that it wouldn't be, though; the closest analogue is Kecleon's Color Collector, and that isn't immutable either.

zinniagamedev commented on PBS/moves_new.txt:
```diff
@@ -1245,6 +1245,20 @@ Flags = CanProtect
+Priority = -6
+FunctionCode = MovesHazardsToOpponentAndForcesSwitch
+Flags = CanProtect,Wind
+Description = A gale clears hazards on your side of the field, moves hazards on the user's side to the opponent's, then forces the targets to switch.
```
This description repeats itself.

bellarch replied:
Oops, typo -- that's supposed to say that it first clears the opponent's side of hazards!

zinniagamedev commented on Plugins/Tectonic Battle/Move/Move_Usage_Calculations.rb:
```diff
@@ -395,6 +395,7 @@ def canApplyRandomAddedEffects?(user, target, showMessages = false, aiCheck = fa
 
     def pbAdditionalEffectChance(user, target, type, effectChance = 0, aiCheck = false)
         return 100 if @battle.pbCheckGlobalAbility(:WISHMAKER)
+        return 100 if user.hasActiveAbility?(:CATASTROPHIC)
```
Don't need a custom line for this. The CertainAddedEffectUserAbility battle handler already covers this type of thing.

zinniagamedev committed:
Tweak the new switchOutUserForSelectedPokemon method

Remove some unused params

zinniagamedev commented on Plugins/Tectonic Battle/Move Codes/Move_Codes_Revival.rb:
```diff
@@ -38,4 +38,55 @@ def effectOnPartyMember(pokemon)
+
+    def effectOnPartyMember(pokemon)
+        pokemon.heal
+        pokemon.hp *= 0.5
```
Need to round the result here! Shouldn't allow "hp" to be a floating point value.

zinniagamedev commented on Plugins/Tectonic Battle/Move Codes/Hazards/Move_Codes_HazardRemoval.rb:
```diff
@@ -171,4 +171,64 @@ def getEffectScore(user, _target)
+        end
+        eachHazard(ourSide, true) do |effect, data|
+            if data.is_hazard?
+                @battle.pbDisplay(_INTL("{1} blew the {2} to the other side of the field!", user.pbThis, data.name)) unless data.has_expire_proc?
```
We might want the two messages to be more distinct? Easy to miss a single preposition change ("on" vs "to"). Not sure.

bellarch replied:
It makes sense to me, and certainly couldn't hurt. How about: "The {1} was blown to the other side of the field!" Or would we want it to be made a bit more distinct than that?

zinniagamedev commented on Plugins/Tectonic Battle/Move Codes/Hazards/Move_Codes_HazardRemoval.rb:
```diff
@@ -171,4 +171,64 @@ def getEffectScore(user, _target)
+            if data.is_hazard?
+                @battle.pbDisplay(_INTL("{1} blew the {2} to the other side of the field!", user.pbThis, data.name)) unless data.has_expire_proc?
+            end
+            if data.is_spike?
```
Should check whether the effect is an integer or a boolean, not whether its a spike. That may not always be the same exact thing forever.

bellarch committed:
Changes based on review

zinniagamedev committed:
FIx CATASTROPHIC CertainAddedEffectUserAbility block

zinniagamedev committed:
Tweak the math on RevivePartyMemberTo50HPThenSwitch's healing

Always round up, to avoid weird issues with mons with 1 HP (Shedinja)
And use totalhp instead of hp for clarity
